proxy/handlers.go

```diff
@@ -223,36 +227,62 @@ func (p *Proxy) OAuthCallback(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, stateParameter.RedirectURI, http.StatusFound)
 }
 
+// shouldSkipAuthentication contains conditions for skipping authentication.
+// Conditions should be few in number and have strong justifications.
+func (p *Proxy) shouldSkipAuthentication(r *http.Request) bool {
+	pol, foundPolicy := p.policy(r)
+
+	if isCORSPreflight(r) && foundPolicy && pol.CORSAllowPreflight {
+		log.FromRequest(r).Debug().Msg("proxy: skipping authentication for valid CORS preflight request")
+		return true
+	}
+
+	return false
+}
+
+// isCORSPreflight inspects the request to see if this is a valid CORS preflight request.
+// These checks are not exhaustive, because the proxied server should be verifying it as well.
+//
+// See https://www.html5rocks.com/static/images/cors_server_flowchart.png for more info.
+func isCORSPreflight(r *http.Request) bool {
+	_, ok1 := r.Header["Access-Control-Request-Method"]
+	_, ok2 := r.Header["Origin"]
```

Comment:
I wonder if it is safer to use r.Header.Get("X") here? 

See : 
- https://tachingchen.com/blog/pitfall-of-golang-header-operation/
- https://golang.org/pkg/net/http/#Header.Get

Reply:
Good call - thanks for the links. Fixed.